test/libraries/SafeToL2Setup.spec.ts

```diff
@@ -0,0 +1,245 @@
+                    safeL2,
+                    proxyFactory,
+                    signers: [user1],
+                    safeToL2SetupLib,
+                } = await setupTests();
+
+                const safeL2SingletonAddress = await safeL2.getAddress();
+                const safeToL2SetupLibAddress = await safeToL2SetupLib.getAddress();
+                const safeToL2SetupCall = safeToL2SetupLib.interface.encodeFunctionData("setupToL2", [safeL2SingletonAddress]);
+
+                const setupData = safeL2.interface.encodeFunctionData("setup", [
+                    [user1.address],
+                    1,
+                    safeToL2SetupLib.target,
+                    safeToL2SetupCall,
+                    ethers.ZeroAddress,
+                    ethers.ZeroAddress,
+                    0,
+                    ethers.ZeroAddress,
+                ]);
+                const safeAddress = await proxyFactory.createProxyWithNonce.staticCall(safeSingleton.target, setupData, 0);
+
+                const transaction = await (await proxyFactory.createProxyWithNonce(safeSingleton.target, setupData, 0)).wait();
+                if (!transaction?.hash) {
+                    throw new Error("No transaction hash");
+                }
+                // I decided to use tracing for this test because it gives an overview of all the storage slots involved in the transaction
+                // Alternatively, one could use `eth_getStorageAt` to check storage slots directly
+                // But that would not guarantee that other storage slots were not touched during the transaction
+                const trace = (await hre.network.provider.send("debug_traceTransaction", [transaction.hash])) as HardhatTrace;
+                // Hardhat uses the most basic struct/opcode logger tracer: https://geth.ethereum.org/docs/developers/evm-tracing/built-in-tracers#struct-opcode-logger
+                // To find the "snapshot" of the storage before the DELEGATECALL into the library, we need to find the first DELEGATECALL opcode calling into the library
+                // To do that, we search for the DELEGATECALL opcode with the stack input pointing to the library address
+                const delegateCallIntoTheLib = trace.structLogs.findIndex(
+                    (log) =>
+                        log.op === "DELEGATECALL" &&
+                        sameHexString(log.stack[log.stack.length - 2], ethers.zeroPadValue(safeToL2SetupLibAddress, 32).slice(2)),
+                );
+                const preDelegateCallStorage = trace.structLogs[delegateCallIntoTheLib].storage;
+
+                // The SafeSetup event is emitted after the Safe is set up
+                // To get the storage snapshot after the Safe is set up, we need to find the LOG2 opcode with the topic input on the stack equal the SafeSetup event signature
+                const SAFE_SETUP_EVENT_SIGNATURE = ethers.id("SafeSetup(address,address[],uint256,address,address)");
```

Comment:
ubernit: Perhaps a slightly nicer way to do this would be with `safeSingleton.interface.getEvent("SafeSetup").topicHash`. Fine as is though - this is low level stuff anyway.

Reply:
I always forget that I can do this. Thanks